contracts/HATVault.sol

```diff
@@ -0,0 +1,770 @@
+import "./tokenlock/TokenLockFactory.sol";
+import "./interfaces/IHATVault.sol";
+import "./interfaces/IRewardController.sol";
+import "./HATVaultsRegistry.sol";
+
+/** @title A Hats.finance vault which holds the funds for a specific project's
+* bug bounties
+* @author Hats.finance
+* @notice The HATVault can be deposited into in a permissionless maner using
+* the vault’s native token. When a bug is submitted and approved, the bounty 
+* is paid out using the funds in the vault. Bounties are paid out as a
+* percentage of the vault. The percentage is set according to the severity of
+* the bug. Vaults have regular safety periods (typically for an hour twice a
+* day) which are time for the committee to make decisions.
+*
+* In addition to the roles defined in the HATVaultsRegistry, every HATVault 
+* has the roles:
+* Committee - The only address which can submit a claim for a bounty payout
+* and set the maximum bounty.
+* User - Anyone can deposit the vault's native token into the vault and 
+* recieve shares for it. Shares represent the user's relative part in the
+* vault, and when a bounty is paid out, users lose part of their deposits
+* (based on percentage paid), but keep their share of the vault.
+* Users also receive rewards for their deposits, which can be claimed at any
+* time.
+* To withdraw previously deposited tokens, a user must first send a withdraw
+* request, and the withdrawal will be made available after a pending period.
+* Withdrawals are not permitted during safety periods or while there is an 
+* active claim for a bounty payout.
+*
+* Bounties are payed out distributed between a few channels, and that 
+* distribution is set upon creation (the hacker gets part in direct transfer,
+* part in vested reward and part in vested HAT token, part gets rewarded to
+* the committee, part gets swapped to HAT token and burned and/or sent to Hats
+* governance).
+*
+* This project is open-source and can be found at:
+* https://github.com/hats-finance/hats-contracts
+*/
+contract HATVault is IHATVault, ERC4626Upgradeable, OwnableUpgradeable, ReentrancyGuardUpgradeable {
+    using SafeERC20 for IERC20;
+
+    struct Claim {
+        bytes32 claimId;
+        address beneficiary;
+        uint256 bountyPercentage;
+        // the address of the committee at the time of the submission, so that this committee will
+        // be paid their share of the bounty in case the committee changes before claim approval
+        address committee;
+        uint256 createdAt;
+        uint256 challengedAt;
+    }
+
+    struct PendingMaxBounty {
+        uint256 maxBounty;
+        uint256 timestamp;
+    }
+
+    uint256 public constant NULL_UINT = type(uint256).max;
+    address public constant NULL_ADDRESS = 0xFFfFfFffFFfffFFfFFfFFFFFffFFFffffFfFFFfF;
+    uint256 public constant HUNDRED_PERCENT = 10000;
+    uint256 public constant MAX_BOUNTY_LIMIT = 9000; // Max bounty can be up to 90%
```

Comment:
there are vaults rn that have a 99% bounty, so we can't have this. or have them lower their bounty.. i think 99% limit is good

Reply:
it is not a matter of opinion :)  We discovered that if you pay out close to 100% of the vault, bad stuff happens with the shares calculation on new deposits. So if we do 99%, then in theory hafter 4 or 5 payouts of 99% we have paid out 99.999999999% of the vault etc. So we set 90% to be save 

This is also a bug in v1, I think, so these 99% vaults are at risk.

Reply:
it was 20 payouts but anyway that's ok :) do we have note of important changes in v2 that we need to inform the projects about?